File: NBE_5_7_2_02TEAM/src/main/java/io/twogether/nbe_5_7_2_02team/chat/service/ChatRoomService.java
package io.twogether.nbe_5_7_2_02team.chat.service;

import static io.twogether.nbe_5_7_2_02team.global.response.error.ErrorCode.CHAT_ROOM_ALREADY_EXISTS;
import static io.twogether.nbe_5_7_2_02team.global.response.error.ErrorCode.CHAT_ROOM_NOT_FOUND;
import static io.twogether.nbe_5_7_2_02team.global.response.error.ErrorCode.NOT_FOUND_POST;

import io.twogether.nbe_5_7_2_02team.chat.dao.ChatRoomRepository;
import io.twogether.nbe_5_7_2_02team.chat.domain.ChatRoom;
import io.twogether.nbe_5_7_2_02team.chat.dto.response.ChatRoomGetResponse;
import io.twogether.nbe_5_7_2_02team.global.exception.ErrorException;
import io.twogether.nbe_5_7_2_02team.post.dao.PostRepository;
import io.twogether.nbe_5_7_2_02team.post.domain.Post;

import lombok.RequiredArgsConstructor;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ChatRoomService {

    private final ChatRoomRepository chatRoomRepository;
    private final PostRepository postRepository;

    @Transactional(readOnly = true)
    public List<ChatRoomGetResponse> getChatRoomList() {
        List<ChatRoom> chatRoomList = chatRoomRepository.findAll();

        return chatRoomList.stream().map(ChatRoomGetResponse::from).toList();
    }

    @Transactional(readOnly = true)
    public ChatRoomGetResponse getChatRoomByPost(Long postId) {
        Post post =
                postRepository
                        .findById(postId)
                        .orElseThrow(() -> new ErrorException(NOT_FOUND_POST));

        ChatRoom chatRoom =
                chatRoomRepository
                        .findByPost(post)
                        .orElseThrow(() -> new ErrorException(CHAT_ROOM_NOT_FOUND));

        return ChatRoomGetResponse.from(chatRoom);
    }

    @Transactional
    public Long createChatroom(Long postId) {
        Post post =
                postRepository
                        .findById(postId)
                        .orElseThrow(() -> new ErrorException(NOT_FOUND_POST));

        chatRoomRepository
                .findByPost(post)
                .ifPresent(
                        chatRoom -> {
                            throw new ErrorException(CHAT_ROOM_ALREADY_EXISTS);
                        });

        return chatRoomRepository.save(ChatRoom.builder().post(post).build()).getId();
    }

    @Transactional
    public void deleteChatroom(Long id) {
        ChatRoom chatRoom = checkChatRoomExists(id);

        chatRoomRepository.delete(chatRoom);
    }

    public ChatRoom checkChatRoomExists(Long id) {
        return chatRoomRepository
                .findById(id)
                .orElseThrow(() -> new ErrorException(CHAT_ROOM_NOT_FOUND));
    }
}

File: NBE_5_7_2_02TEAM/src/main/java/io/twogether/nbe_5_7_2_02team/chat/service/ChatRoomService.kt
package io.twogether.nbe_5_7_2_02team.chat.service

import io.twogether.nbe_5_7_2_02team.chat.dao.ChatRoomRepository
import io.twogether.nbe_5_7_2_02team.chat.domain.ChatRoom
import io.twogether.nbe_5_7_2_02team.chat.dto.response.ChatRoomGetResponse
import io.twogether.nbe_5_7_2_02team.chat.dto.response.toGetResponse
import io.twogether.nbe_5_7_2_02team.global.exception.ErrorException
import io.twogether.nbe_5_7_2_02team.global.response.error.ErrorCode
import io.twogether.nbe_5_7_2_02team.post.dao.PostRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.function.Supplier

@Service
class ChatRoomService(
    private val chatRoomRepository: ChatRoomRepository,
    private val postRepository: PostRepository,
) {
    @Transactional(readOnly = true)
    fun getChatRoomList(): List<ChatRoomGetResponse> {
        val chatRoomList = chatRoomRepository.findAll()

        return chatRoomList.map { chatRoom -> chatRoom.toGetResponse() }
    }

    @Transactional(readOnly = true)
    fun getChatRoomByPost(postId: Long): ChatRoomGetResponse {
        val post =
            postRepository
                .findById(postId)
                .orElseThrow(Supplier { ErrorException(ErrorCode.NOT_FOUND_POST) })

        val chatRoom =
            chatRoomRepository
                .findByPost(post) ?: throw ErrorException(ErrorCode.CHAT_ROOM_NOT_FOUND)

        return chatRoom.toGetResponse()
    }

    @Transactional
    fun createChatroom(postId: Long): Long {
        val post =
            postRepository
                .findById(postId)
                .orElseThrow(Supplier { ErrorException(ErrorCode.NOT_FOUND_POST) })

        chatRoomRepository
            .findByPost(post)
            ?.let { throw ErrorException(ErrorCode.CHAT_ROOM_ALREADY_EXISTS) }

        return chatRoomRepository.save(ChatRoom(post = post)).id!!
    }

    @Transactional
    fun deleteChatroom(id: Long) {
        val chatRoom = checkChatRoomExists(id)

        chatRoomRepository.delete(chatRoom)
    }

    fun checkChatRoomExists(id: Long): ChatRoom =
        chatRoomRepository.findChatRoomById(id) ?: throw ErrorException(ErrorCode.CHAT_ROOM_NOT_FOUND)
}
